Replace with kfctl_istio_dex and doc improvements for the config

krishnadurai opened the pull request:
Addresses: #1592 

/assign @jlewi @yanniszark

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/website/1707)
<!-- Reviewable:end -->

yanniszark commented:
Thanks @krishnadurai, I'm on it.

yanniszark commented:
@krishnadurai overall seems good.
I only propose one addition.

Many people may have bookmarked the existing_arrikto page and we should give them some info instead of breaking the page.
It would be useful to keep the page with a little text, explaining the transition to a community-supported  config.

I have created a review branch to demonstrate what I mean: https://github.com/yanniszark/website/tree/kfctl_istio_dex-rev
https://github.com/yanniszark/website/commit/aaf2cb8b64bbedc3f482735a71406b297d8c0137
If you're ok with it, just cherry-pick the commit and we should be good to go.

krishnadurai commented:
Hello @yanniszark,
Thanks for your review.

Regarding the placeholder for the erstwhile existing-arrikto config, it may not be necessary for us to add one as users will get a list of supported configurations from the getting started page.

yanniszark commented:
> Regarding the placeholder for the erstwhile existing-arrikto config, it may not be necessary for us to add one as users will get a list of supported configurations from the getting started page.

Many people have it bookmarked, so they won't pass through the getting started page.
In my experience, it's always good practice to leave a notice when something is changed.
In this case, instead of the page disappearing, users will get a nice text explaining what happened and where to go.

cc @sarahmaddox for the docs experience

johnugeorge commented:
@yanniszark The text in your link is not technical and promotes a single vendor. we have discussed this earlier.  The Dex config is meant to be vendor neutral without any vendor references.

jlewi commented:
Is it possible to setup a redirect from
https://www.kubeflow.org/docs/started/k8s/kfctl-existing-arrikto/
to the new docs?

https://github.com/kubeflow/website/blob/master/content/_redirects

yanniszark commented:
@johnugeorge thanks for voicing your concerns, however I don't see any promotion is happening here, the contrary.

This is a transition that we all agreed upon at the relevant issue (https://github.com/kubeflow/manifests/issues/599) and the extra section serves to inform Kubeflow users of what happened with the existing_arrikto config. It is important for users to understand what happened, so they are not lost, and go straight to the new config. Just redirecting will confuse them I think.

We understand and agree that the kfctl_istio_dex config is a vendor-neutral config, that's why the transition happened in the first place.

In any case after your feedback, I have amended the message to make it crystal clear that the new config is a community config and developed/maintained by everyone.
See: https://github.com/yanniszark/website/commit/528242c8fd9daec9d81322e2b900c901fb760900

krishnadurai commented:
@jlewi I'll set up a redirect.

@yanniszark the text which you suggest is conflicting to what was discussed. Contributions to manifests have been from multiple community members and a kfdef is a resultant of it.

yanniszark commented:
> @yanniszark the text which you suggest is conflicting to what was discussed.

I believe the message makes it clear that the new kfctl_istio_dex config is developed by the whole Kubeflow community.

We want to expose to users that this transition happened (as described by issue https://github.com/kubeflow/manifests/issues/599) so they don't get lost.

Can you propose a wording to describe this transition?

sarahmaddox commented:
Previews:
* Redirect works: https://deploy-preview-1707--competent-brattain-de2d6d.netlify.com/docs/started/k8s/kfctl-existing-arrikto/
* New page: https://deploy-preview-1707--competent-brattain-de2d6d.netlify.com/docs/started/k8s/kfctl-istio-dex/
* https://deploy-preview-1707--competent-brattain-de2d6d.netlify.com/docs/started/k8s/overview/

sarahmaddox commented:
/assign

sarahmaddox commented on content/docs/started/k8s/kfctl-k8s-istio.md:
```diff
@@ -77,6 +77,8 @@ Notes:
   customize if necessary.
 
+* **Disabling Istio Installation** - In the configuration file {{% config-file-istio-dex %}} you may choose to not install Istio by removing the applications `istio-crds` and `istio-install`, if your Kubernetes cluster has an existing Istio installation.
+
```
While you're here, please would you remove the following line from the top of this file: **Maintainer and supporter: Kubeflow community**. That line is no longer relevant, because both configs are supported by the Kubeflow community.

krishnadurai replied:
Sure

sarahmaddox commented on content/docs/started/k8s/kfctl-istio-dex.md:
```diff
@@ -78,11 +77,15 @@ Notes:
   a local version of the configuration YAML file which you can further
   customize if necessary.
 
+* **Disabling Istio Installation** - In the configuration file {{% config-file-istio-dex %}} you may choose to not install Istio by removing the applications `istio-crds` and `istio-install`, if your Kubernetes cluster has an existing Istio installation.
```
s/Installation/installation

(The Kubeflow style guide recommends that we avoid capital letters unless we're referring to a product or brand name. Capital letters can be confusing if used for non-brand words.)

sarahmaddox commented on content/docs/started/k8s/kfctl-istio-dex.md:
```diff
@@ -78,11 +77,15 @@ Notes:
 
+* **Disabling Istio Installation** - In the configuration file {{% config-file-istio-dex %}} you may choose to not install Istio by removing the applications `istio-crds` and `istio-install`, if your Kubernetes cluster has an existing Istio installation.
+
+* **Default Static File Password** - The configuration file {{% config-file-istio-dex %}} contains a default [staticPasswords](https://github.com/dexidp/dex/blob/0f8c4db9f61476a8f80e60f5950992149a1cc0cb/examples/config-dev.yaml#L91-L95) user with email set to `admin@kubeflow.org` and password `12341234`. It is encouraged to change this configuration or replace it with a [Dex connector](https://gi ...
```
Remove capital letters and clarify the heading: `Default password in static file`.

Also for clarity about who should take action, change `It is encouraged to change...` to `You should change...`.

sarahmaddox commented on content/docs/started/k8s/kfctl-istio-dex.md:
```diff
@@ -1,21 +1,20 @@
-[Arrikto](https://www.arrikto.com/), it is geared towards existing Kubernetes
+This installation of Kubeflow is geared towards existing Kubernetes
 clusters and does not depend on any cloud-specific feature.
 
```
Add a heading here (immediately after the new note suggested in my previous comment) to break up the long introduction:

```
## Overview
```

krishnadurai replied:
Sure I'll include this header.

krishnadurai replied:
The word overview alone isn't rendering properly on the website. I've chosen to go with Architecture Overview instead

sarahmaddox replied:
Ah yes, I've seen that before. Strange!

sarahmaddox commented on content/docs/started/k8s/kfctl-istio-dex.md:
```diff
@@ -1,21 +1,20 @@
-This installation of Kubeflow is maintained by
-[Arrikto](https://www.arrikto.com/), it is geared towards existing Kubernetes
+This installation of Kubeflow is geared towards existing Kubernetes
 clusters and does not depend on any cloud-specific feature.
```
+1 to the comment from @yanniszark that we should help existing users understand that there's a transition from the previous config in v0.7 and the new config in v1.0. I can't find the previous suggested wording. I think we need to add a note at the top of this file, and keep the note as simple as possible. 

Also +1 to the redirect suggested by @jlewi and @krishnadurai in this PR. 

I'm not completely aware of all the complexities, so please excuse me if I inadvertently tread on any toes. Does the following note work, for adding after line 10?

```
**Note:** This installation uses the `kfctl_istio_dex` configuration (configuration file {{% config-uri-istio-dex %}}). This configuration replaces the `kfctl_existing_arrikto` configuration that was used in Kubeflow v0.7 and earlier.
```

krishnadurai replied:
@sarahmaddox Wouldn't it be simpler to just mention that kfctl_existing_arrikto is no longer supported?

sarahmaddox replied:
@krishnadurai I think it's clearer to tell people that this config replaces the earlier one. The current wording seems out of place, because it's talking about a config that's not mentioned anywhere else. People will wonder why we're mentioning it:
`Note: The configuration kfctl_existing_arrikto that was used in Kubeflow v0.7 and earlier is no longer supported.`

The wording in my earlier comment mentions that the new config is a replacement for the earlier one.

In a later comment, @yanniszark has said he's OK with my suggested wording too. I recommend we go with it.

johnugeorge commented:
> @johnugeorge thanks for voicing your concerns, however I don't see any promotion is happening here, the contrary.
> 
> This is a transition that we all agreed upon at the relevant issue ([kubeflow/manifests#599](https://github.com/kubeflow/manifests/issues/599)) and the extra section serves to inform Kubeflow users of what happened with the existing_arrikto config. It is important for users to understand what happened, so they are not lost, and go straight to the new config. Just redirecting will confuse them I think.
> 
> We understand and agree that the kfctl_istio_dex config is a vendor-neutral config, that's why the transition happened in the first place.
> 
> In any case after your feedback, I have amended the message to make it crystal clear that the new config is a community config and developed/maintained by everyone.
> See: [yanniszark@528242c](https://github.com/yanniszark/website/commit/528242c8fd9daec9d81322e2b900c901fb760900)

@yanniszark Thanks for your view point. I want to raise some factual inaccuracies here. Transition of existing_arrikto config(or origin of https://github.com/kubeflow/manifests/issues/599)  is a result of the discussion happened in https://github.com/kubeflow/manifests/pull/560.  From the community point of view, the right thing to do always is to have clear distinction between vendor specific and community supported configs. https://github.com/kubeflow/manifests/pull/560 by @krishnadurai  was aimed at having a vendor neutral dex config for the community since there was none available at that time. However in the discussion held later, it was conveyed that vendor specific existing_arrikto config is not planned to be maintained separately in the future and it can be converged to a community config together with changes in the PR(https://github.com/kubeflow/manifests/pull/560#issuecomment-548849626). 

Given this context, i believe that it is incorrect to say  "After a lot of requests by the community, we turned the Arrikto config into a community-supported config" or give any references to existing_arrikto config (as decided before). It has contributions from multiple companies. As from the earlier comments, we could either add a redirect or give a warning that it is not supported anymore(and point to new url).

krishnadurai commented on content/docs/started/k8s/overview.md:
```diff
@@ -119,27 +119,10 @@ This section includes vendor-neutral solutions governed by community consensus.
-  </table>
-</div>
-
-### Vendor maintained
```
This makes sense. I'll include these changes.

sarahmaddox replied:
Thanks!

krishnadurai commented on content/docs/started/k8s/kfctl-k8s-istio.md:
```diff
@@ -77,6 +77,8 @@ Notes:
   a local version of the configuration YAML file which you can further
   customize if necessary.
 
+* **Disabling Istio Installation** - In the configuration file {{% config-file-istio-dex %}} you may choose to not install Istio by removing the applications `istio-crds` and `istio-install`, if your Kubernetes cluster has an existing Istio installation.
```
@sarahmaddox its a copy-paste error which I have overlooked. The config file should have been k8s-istio instead of istio-dex. I'll correct it here.

sarahmaddox replied:
Thanks!

NikeNano commented on content/docs/started/k8s/kfctl-istio-dex.md:
```diff
@@ -1,21 +1,20 @@
-![kfctl_existing_arrikto_architecture](../../kfctl_existing_arrikto-architecture.svg)
+![kfctl_istio_dex_architecture](../../kfctl_istio_dex-architecture.svg)
 
 Read the relevant [article](https://journal.arrikto.com/kubeflow-authentication-with-istio-dex-5eafdfac4782) for more info about this architecture.
```
Will this link also be relevant after the updates?

krishnadurai replied:
Hello @NikeNano, I think it's relevant as it describes the approach taken for this configuration.

NikeNano replied:
Sounds good,I was not sure how big the differences would be between this and the original arrikto configurations.

yanniszark commented:
@johnugeorge we both agree that kfctl_istio_dex is a community-supported config. We also agree that users coming from v0.7 kfctl_existing_arrikto should use kfctl_istio_dex, because it is the virtually the same configuration and allows for a smooth transition.

It would be incorrect to say existing_arrikto is not supported, as it would misrepresent our intention in performing this migration. Sarah's proposal is a good compromise:

> **Note:** This installation uses the `kfctl_istio_dex` configuration (configuration file {{% config-uri-istio-dex %}}). This configuration replaces the `kfctl_existing_arrikto` configuration that was used in Kubeflow v0.7 and earlier.

sarahmaddox commented:
Thanks very much @krishnadurai I think there's just one comment to resolve - the note about the previous config.

krishnadurai commented:
Hello @sarahmaddox and @yanniszark,

I've addressed your concerns by including the note with the same suggested wording. I have placed the note in kfctl_existing_arrikto page – which now wouldn’t redirect to kfctl_istio_dex.

There are a couple of advantages to this approach:
 
1. A user looking for kfctl_existing_arrikto configuration lands on the intended page (/docs/started/k8s/kfctl-existing-arrikto/) and sees the notice about which configuration to use. Using the notice's hyperlink, they can navigate to kfctl_istio_dex configuration.
 
2. A user discovers kfctl_istio_dex configuration lands on the intended page (/docs/started/k8s/kfctl-istio-dex/) and need not be aware of any note and has a smoother user experience.

This approach is better for both sets of users as each of them gets their relevant information to install Kubeflow.

yanniszark commented:
@krishnadurai thanks, text looks good and should inform users.
I'm not sure if it's better to include a note or not, since now the existing_arrikto page is hidden and cannot be navigated to.
@sarahmaddox what do you think is better for UX?

sarahmaddox commented on layouts/partials/sidebar-tree.html:
```diff
@@ -30,7 +30,7 @@
   <ul>
     <li class="collapse {{ if $show }}show{{ end }}" id="{{ $sid }}">
-      {{ $pages := where (union $s.Pages $s.Sections).ByWeight ".Params.toc_hide" "!=" true }}
+      {{ $pages := where (union $s.Pages $s.Sections).ByWeight ".Params.page_hide" "!=" true }}
```
I think we should revert this change. It's a confusing user experience to land on a page that isn't linked in the nav. If the page exists, we should show it in the nav and in the [landing page](https://deploy-preview-1707--competent-brattain-de2d6d.netlify.com/docs/started/k8s/). The page is giving useful information at this stage. When we no longer need the page, we will remove it.

krishnadurai replied:
Sure @sarahmaddox, I'll revert this change.

I've made an issue to track the removal of this page: kubeflow/website#738

sarahmaddox replied:
Sounds good, thanks!

sarahmaddox commented on themes/docsy/layouts/partials/section-index.html:
```diff
@@ -16,7 +16,7 @@
     <hr class="panel-line">
         {{ range $pages }}
-            {{ if eq .Parent $parent }}
+            {{ if and (eq .Parent $parent) (ne .Params.page_hide true) }}
```
Revert this change. We should not change the files in the `/themes` directory, as any changes here will be overridden next time we update our version of Docsy.

sarahmaddox commented on themes/docsy/layouts/partials/sidebar-tree.html:
```diff
@@ -30,7 +30,7 @@
   <ul>
     <li class="collapse {{ if $show }}show{{ end }}" id="{{ $sid }}">
-      {{ $pages := where (union $s.Pages $s.Sections).ByWeight ".Params.toc_hide" "!=" true }}
+      {{ $pages := where (union $s.Pages $s.Sections).ByWeight ".Params.page_hide" "!=" true }}
```
Revert this change. We should not change the files in the `/themes` directory, as any changes here will be overridden next time we update our version of Docsy. This file is overridden by the one in the main `/layouts` directory.

sarahmaddox commented on content/docs/started/k8s/kfctl-existing-arrikto.md:
```diff
@@ -0,0 +1,8 @@
+page_hide = true
++++
+
+If you were using the `kfctl_existing_arrikto` configuration in Kubeflow v0.7 or earlier, you should use {{% config-uri-istio-dex %}} in Kubeflow {{% kf-latest-version %}} following the instructions in the [`kfctl_istio_dex`](/docs/started/k8s/kfctl-istio-dex/) page.
```
How about this slight change, for more clarity:

```
If you were using the `kfctl_existing_arrikto` configuration in Kubeflow v0.7 or earlier, you should use {{% config-file-istio-dex %}} in Kubeflow {{% kf-latest-version %}}. Follow the instructions in the [guide to `kfctl_istio_dex`](/docs/started/k8s/kfctl-istio-dex/).
```

(I've changed the wording of the last part of the sentence, and also change the URI to just the file name in the variable.)

krishnadurai replied:
Sure

sarahmaddox replied:
Thanks!

krishnadurai commented on content/docs/started/k8s/kfctl-istio-dex.md:
```diff
@@ -0,0 +1,853 @@
+* **Istio configuration for trustworthy JWTs** - This configuration uses Istio version
+  1.3.1 with SDS enabled, which requires you to use Kubernetes 1.13 or later. You may need
+  to add extra configurations to your Kubernetes as mentioned in
+  [Istio's blog](https://istio.io/blog/2019/trustworthy-jwt-sds/).
```
@sarahmaddox I've added this additional note for installation. Please review it.

sarahmaddox replied:
Thanks, LGTM.

sarahmaddox commented:
Thanks @krishnadurai, and thanks to everyone who's contributed feedback to this PR.

/lgtm
/approve
